File: aricode/sparring/challenges/12_perceptron/perceptron.c
/* C - Challenge 12: Perceptron Neural Network */
/* Train AND gate, 1000 epochs, fixed-point (scale=1000) */
/* Return correct predictions (expect 4) */

int train_and_predict(void) {
    int w1 = 0, w2 = 0, bias = 0;
    int lr = 100;

    for (int epoch = 0; epoch < 1000; epoch++) {
        int out, pred, err;

        /* (0,0) -> 0 */
        out = bias; pred = out > 500 ? 1000 : 0;
        err = 0 - pred;
        bias += lr * err / 1000;

        /* (0,1) -> 0 */
        out = w2 + bias; pred = out > 500 ? 1000 : 0;
        err = 0 - pred;
        w2 += lr * err / 1000; bias += lr * err / 1000;

        /* (1,0) -> 0 */
        out = w1 + bias; pred = out > 500 ? 1000 : 0;
        err = 0 - pred;
        w1 += lr * err / 1000; bias += lr * err / 1000;

        /* (1,1) -> 1 */
        out = w1 + w2 + bias; pred = out > 500 ? 1000 : 0;
        err = 1000 - pred;
        w1 += lr * err / 1000; w2 += lr * err / 1000; bias += lr * err / 1000;
    }

    int correct = 0;
    if (bias <= 500) correct++;
    if (w2 + bias <= 500) correct++;
    if (w1 + bias <= 500) correct++;
    if (w1 + w2 + bias > 500) correct++;
    return correct;
}

int main(void) { return train_and_predict(); }

File: aricode/sparring/challenges/12_perceptron/perceptron.rs
// Rust - Challenge 12: Perceptron Neural Network
// Train AND gate, 1000 epochs, fixed-point, return correct (4)
use std::process::ExitCode;

fn train() -> i32 {
    let (mut w1, mut w2, mut bias) = (0i32, 0i32, 0i32);
    let lr = 100;
    let samples = [(0,0,0), (0,1000,0), (1000,0,0), (1000,1000,1000)];
    for _ in 0..1000 {
        for &(x1, x2, target) in &samples {
            let out = w1*x1/1000 + w2*x2/1000 + bias;
            let pred = if out > 500 { 1000 } else { 0 };
            let err = target - pred;
            w1 += lr * err * x1 / 1000 / 1000;
            w2 += lr * err * x2 / 1000 / 1000;
            bias += lr * err / 1000;
        }
    }
    let mut c = 0;
    if bias <= 500 { c += 1; }
    if w2+bias <= 500 { c += 1; }
    if w1+bias <= 500 { c += 1; }
    if w1+w2+bias > 500 { c += 1; }
    c
}

fn main() -> ExitCode { ExitCode::from(train() as u8) }
